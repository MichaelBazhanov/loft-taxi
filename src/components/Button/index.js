export { default } from "./Button";
export { styles } from "./styles";
