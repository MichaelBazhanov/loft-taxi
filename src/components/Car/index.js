export { default } from "./Car";
export { styles } from "./styles";
