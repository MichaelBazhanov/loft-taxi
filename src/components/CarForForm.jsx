import React from "react";
import PropTypes from "prop-types";

function classNames(...classes) {
	return classes.filter(Boolean).join(' ')
}

const CarForForm = ({ price, imgSRC, index, activeIndexCar, setActiveIndexCar }) => {
	const toggleActive = () => {
		setActiveIndexCar(index)
	}

	return (
		<div className={classNames(index === activeIndexCar ? '' : 'opacity-50', 'w-1/3 px-3 cursor-pointer')} onClick={toggleActive}>
			<div className="flex flex-col shadow-me rounded-lg p-3">
				<p>Стандарт</p>
				<p className="text-xs">Стоимость</p>
				<p className="text-2xl">{price}</p>
				<img src={imgSRC} alt="car" className="w-full h-auto" />
			</div>
		</div>
	)
}

CarForForm.propTypes = {
	price: PropTypes.string,
	imgSRC: PropTypes.string,
	index: PropTypes.number,
	activeIndexCar: PropTypes.number,
	setActiveIndexCar: PropTypes.func
}

export default CarForForm

