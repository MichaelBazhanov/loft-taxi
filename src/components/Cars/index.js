export { default } from "./Cars";
export { styles } from "./styles";
