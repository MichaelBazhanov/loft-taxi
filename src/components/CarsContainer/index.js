export { default } from "./CarsContainer";
export { styles } from "./styles";
