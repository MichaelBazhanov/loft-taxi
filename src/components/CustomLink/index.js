export { default } from './CustomLink' // экспорт по умолчанию