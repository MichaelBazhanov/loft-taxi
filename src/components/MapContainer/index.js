export { HOCMapContainer, ContextMap } from "./MapContainer";
export { styles } from "./styles";
