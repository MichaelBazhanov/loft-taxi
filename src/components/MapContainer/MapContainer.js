import React, { useState, useEffect } from "react";
import { connect } from "react-redux";
import {
  getRoutesCoordinates,
  resetRoutesAndAddress,
} from "../../modules/route";
import { getPaymentCard } from "../../modules/payment";
import { getAddressList } from "../../modules/address";
import Select from "../Select";
import { useNavigate } from "react-router-dom";

//img
import imgStandart from "../../assets/images/car-standart.jpg";
import imgPremium from "../../assets/images/car-premium.jpg";
import imgBusiness from "../../assets/images/car-business.jpg";

import Loading from "../Loading";
import Error from "../Error";

//Car
import CarForForm from "../CarForForm";

//=========================================
import FormSelectors from "../FormSelectors";
//=========================================

function classNames(...classes) {
  return classes.filter(Boolean).join(" ");
}

const FormForMap = ({
  getAddressList,
  getRoutesCoordinates,
  resetRoutesAndAddress,
  getPaymentCard,
  isLoadingSendPaymentCardNewUser,
  errorSendPaymentCardNewUser,
  isLoading,
  error,
  address,
  cardName,
  cardNumber,
  expiryDate,
  cvc,
  token,
}) => {
  const [width, setWidth] = useState(window.innerWidth);
  const [activeBlock, setActiveBlock] = useState("without-card"); //1 = 'without-card', 2 = 'next-order', 3 = 'default'
  const [activeIndexCar, setActiveIndexCar] = useState(1);
  const [addressStart, setAddressStart] = useState(null);
  const [addressEnd, setAddressEnd] = useState(null);

  const navigate = useNavigate();

  const updateDimensions = () => {
    setWidth(window.innerWidth);
  };
  useEffect(() => {
    window.addEventListener("resize", updateDimensions);
    return () => window.removeEventListener("resize", updateDimensions);
  }, []);

  const handleSubmit = (event) => {
    event.preventDefault();
    setActiveBlock("next-order"); // Следующий заказ

    if (addressStart && addressEnd && addressStart.rout && addressEnd.rout) {
      getRoutesCoordinates(addressStart, addressEnd);
    }
  };

  //Все useEffect при первом рендере выполняются !!!
  useEffect(() => {
    getAddressList();
  }, []);

  useEffect(() => {
    if (!isLoadingSendPaymentCardNewUser) getPaymentCard(token); // Если для нового пользователя карта уже установлена то ...
  }, [isLoadingSendPaymentCardNewUser]);

  useEffect(() => {
    if (
      cardName === "" ||
      cardName === " " ||
      cardNumber === "" ||
      cardNumber === " " ||
      expiryDate === "" ||
      expiryDate === " " ||
      cvc === "" ||
      cvc === " "
    ) {
      setActiveBlock("default"); // Данные карты пустые!
    } else {
      setActiveBlock("without-card"); // Данные карты НЕ пустые!
    }
  }, [cardName, cardNumber, expiryDate, cvc]);

  if (isLoading) return <Loading />;
  if (error) return <Error />;
  if (isLoadingSendPaymentCardNewUser) return <Loading />;
  if (errorSendPaymentCardNewUser) return <Error />;

  return (
    <div className="container mx-auto relative pointer-events-none h-full">
      <div className="flex flex-col items-center lg:items-stretch h-full">

        {/* Форма */}
        {activeBlock === "without-card" && (
          <div className="flex flex-col h-full">
            <form
              onSubmit={handleSubmit}
              className="sm:max-w-[486px]  w-full sm:mt-16 lg:ml-24 rounded-xl sm:shadow-lg
              flex flex-col h-full sm:h-auto sm:bg-white
              "
            >
              <FormSelectors
                className={
                  width > 640
                    ? "p-6 pb-0 bg-white pointer-events-auto shadow-lg sm:shadow-none"
                    : "bg-white pointer-events-auto shadow-lg sm:shadow-none p-3 pb-0 sm:hidden mt-10 rounded-xl w-full"
                }
                margin={width > 640 ? "mt-1" : "mt-0"}
                placeholders={["Откуда", "Куда"]}
                address={address}
                addressStart={addressStart}
                addressEnd={addressEnd}
                setAddressStart={setAddressStart}
                setAddressEnd={setAddressEnd}
              />

              {/* Машинки и заказать */}
              <div className="py-4 px-3 sm:py-8 sm:px-10 sm:mt-6 rounded-xl border mt-auto bg-white shadow-lg sm:shadow-none pointer-events-auto">
                <div className="flex -mx-3">
                  <CarForForm
                    price={"150 ₽"}
                    imgSRC={imgStandart}
                    index={1}
                    setActiveIndexCar={setActiveIndexCar}
                    activeIndexCar={activeIndexCar}
                  />
                  <CarForForm
                    price={"200 ₽"}
                    imgSRC={imgPremium}
                    index={2}
                    setActiveIndexCar={setActiveIndexCar}
                    activeIndexCar={activeIndexCar}
                  />
                  <CarForForm
                    price={"300 ₽"}
                    imgSRC={imgBusiness}
                    index={3}
                    setActiveIndexCar={setActiveIndexCar}
                    activeIndexCar={activeIndexCar}
                  />
                </div>

                <button
                  type="submit"
                  className="text-2xl py-4 w-full bg-yellow-me rounded-full mt-7 disabled:opacity-75"
                  disabled={
                    !(
                      addressStart &&
                      addressStart.rout &&
                      addressEnd &&
                      addressEnd.rout
                    )
                  }
                >
                  Заказать
                </button>
              </div>
            </form>
          </div>
        )}

        {/* Заказ размещен */}
        {activeBlock === "next-order" && (
          <div
            className={classNames(
              width < 1024 ? "mt-auto" : "",
              "max-w-[486px] w-full bg-white lg:mt-16 xl:ml-24 rounded-xl shadow-lg p-3 lg:py-10 lg:px-11 pointer-events-auto text-center lg:text-left"
            )}
          >
            <p className="font-bold text-xl lg:text-4xl">Заказ размещен</p>
            <p className="mt-3 xl:mt-4 text-base lg:text-lg text-gray-me">
              Ваше такси уже едет к вам. Прибудет приблизительно через 10 минут.
            </p>
            <button
              onClick={() => {
                resetRoutesAndAddress(); //обнуляем в redux
                setActiveBlock("without-card"); // активный индекс "блока"
                setAddressStart(null); //обнуляем state
                setAddressEnd(null); //обнуляем state
                setActiveIndexCar(1); // устанавливаем первый индекс
              }}
              type="button"
              className="text-lg lg:text-2xl py-2 lg:py-4 w-full bg-yellow-me rounded-full mt-2 lg:mt-7"
            >
              Сделать новый заказ
            </button>
          </div>
        )}

        {/* Заполните платежные данные */}
        {activeBlock === "default" && (
          <div
            className={classNames(
              width < 640 ? "mt-auto" : "",
              "max-w-[486px] w-full bg-white sm:mt-16 sm:ml-24 rounded-xl shadow-lg p-3 sm:py-10 sm:px-11 pointer-events-auto text-center sm:text-left"
            )}
          >
            <p className="font-bold text-xl sm:text-4xl">
              Заполните платежные данные
            </p>
            <p className="mt-3 sm:mt-4 text-base sm:text-lg text-gray-me">
              Укажите информацию о платежной карте что бы сделать заказ.
            </p>
            <button
              onClick={() => {
                navigate("/profile");
              }}
              type="button"
              className="text-lg sm:text-2xl py-2 sm:py-4 w-full bg-yellow-me rounded-full mt-2 sm:mt-7"
            >
              Перейти в профиль
            </button>
          </div>
        )}
      </div>
    </div>
  );
};
export default connect(
  (state) => ({
    token: state.authorizationReducer.token,

    isLoading: state.addressReducer.isLoading,
    error: state.addressReducer.error,
    address: state.addressReducer.address,

    cardName: state.paymentReducer.cardName,
    cardNumber: state.paymentReducer.cardNumber,
    expiryDate: state.paymentReducer.expiryDate,
    cvc: state.paymentReducer.cvc,

    isLoadingSendPaymentCardNewUser:
      state.paymentReducer.isLoadingSendPaymentCardNewUser,
    errorSendPaymentCardNewUser:
      state.paymentReducer.errorSendPaymentCardNewUser,
  }),
  {
    getAddressList,
    getRoutesCoordinates,
    getPaymentCard,
    resetRoutesAndAddress,
  } // просто дергаем ACTION
)(FormForMap);
