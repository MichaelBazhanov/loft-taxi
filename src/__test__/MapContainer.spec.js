import React from "react";
import TestRenderer from "react-test-renderer";
import { Provider } from "react-redux";

import { HOCMapContainer } from "../components/MapContainer";

// Мокаем useNavigate что бы внутри <HOCMapContainer /> отдал просто функцию
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => jest.fn(),
}));
// Формируем замоканый стор
const mockStore = {
  getState: () => ({
    // входящие пропсы
    authorizationReducer: { token: "123" },

    routesReducer: {
      coordinates: [1,2],
    },

    addressReducer: {
      address: [
        { id: 1, root: "Адрес 1" },
        { id: 2, root: "Адрес 2" },
      ],
      isLoading: false,
      error: false,
    },
    paymentReducer: {
      cardName: "test",
      cardNumber: "1",
      expiryDate: "000",
      cvc: "1",
      isLoadingSendPaymentCardNewUser: false,
      errorSendPaymentCardNewUser: false,
    },
  }),
  subscribe: () => {},
  dispatch: () => {
    // диспачим внутри компонента
    getAddressList: jest.fn();
    getRoutesCoordinates: jest.fn();
    getPaymentCard: jest.fn();
  },
};

it("Renders correctly", () => {
  const FormForMapExample = TestRenderer.create(
    <Provider store={mockStore}>
      <HOCMapContainer />
    </Provider>
  ).toJSON();

  expect(FormForMapExample).toMatchSnapshot();
});
