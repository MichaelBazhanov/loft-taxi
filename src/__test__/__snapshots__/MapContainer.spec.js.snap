// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<div
  className="container mx-auto relative pointer-events-none h-full"
>
  <div
    className="flex flex-col items-center lg:items-stretch h-full"
  >
    <div
      className="max-w-[486px] w-full mt-auto lg:mt-16 xl:ml-24 lg:py-10 lg:px-11 bg-white lg:text-left rounded-xl shadow-lg p-3 pointer-events-auto text-center"
    >
      <p
        className="font-bold text-xl lg:text-4xl"
      >
        Заказ размещен
      </p>
      <p
        className="mt-3 xl:mt-4 text-base lg:text-lg text-gray-me"
      >
        Ваше такси уже едет к вам. Прибудет приблизительно через 10 минут.
      </p>
      <button
        className="w-full py-2 mt-2 lg:mt-7 lg:py-4 text-lg lg:text-2xl bg-yellow-me rounded-full"
        onClick={[Function]}
        type="button"
      >
        Сделать новый заказ
      </button>
    </div>
  </div>
</div>
`;
